lib/internal/test_runner/test.js

```diff
@@ -362,8 +364,37 @@ class Test extends AsyncResource {
   }
 
   matchesTestNamePatterns() {
-    return ArrayPrototypeSome(testNamePatterns, (re) => RegExpPrototypeExec(re, this.name) !== null) ||
-      this.parent?.matchesTestNamePatterns();
+    const matchesByNameOrParent = ArrayPrototypeSome(testNamePatterns, (re) =>
+      RegExpPrototypeExec(re, this.name) !== null ||
+      this.parent?.matchesTestNamePatterns(),
+    );
+
+    if (matchesByNameOrParent) return true;
+
+    const testNameWithAncestors = this.getTestNameWithAncestors();
+    if (!testNameWithAncestors) return false;
+
+    return ArrayPrototypeSome(testNamePatterns, (re) => RegExpPrototypeExec(re, testNameWithAncestors) !== null);
+  }
+
+  /**
+   * Returns a name of the test prefixed by name of all its ancestors in ascending order, separated by a space
+   * Ex."grandparent parent test"
+   *
+   * It's needed to match a single test with non-unique name by pattern
+   */
+  getTestNameWithAncestors() {
+    if (!this.nesting) return;
+
+    const ancestorNames = Array(this.nesting);
+    let parent = this.parent;
+
+    for (let i = 0; i < this.nesting; i++) {
+      ancestorNames[this.nesting - i - 1] = parent.name;
+      parent = parent.parent;
+    }
+
+    return `${ArrayPrototypeJoin(ancestorNames, ' ')} ${this.name}`;
   }
```

Comment:
```suggestion
  getTestNameWithAncestors() {
    if (!this.parent) return '';
   return `${this.parent.getTestNameWithAncestors()} ${this.name}`;
  }
```

Reply:
hm, recursion instead of iteration - ok, I will test it

Reply:
ok, it works the same but because result of this function is always prefixed with empty space I must trim it also like

```js
const testNameWithAncestors = StringPrototypeTrim(this.getTestNameWithAncestors());
```

sounds ok?

tests pass with this

Reply:
yes! recursion just seems more readable to me in this case